fedimint-atomic-broadcast/src/broadcast.rs

```diff
@@ -1,155 +1,175 @@
+        outgoing_message_sender: async_channel::Sender<(Message, Recipient)>,
+    ) -> Self {
+        let (network_data_sender, network_data_receiver) = async_channel::bounded(256);
+        let (signed_block_sender, signed_block_receiver) = async_channel::bounded(16);
+
+        let db_clone = db.clone();
+        let sender_clone = outgoing_message_sender.clone();
+
+        let relay_handle = tokio::spawn(async move {
+            while let Ok((message, peer_id)) = incoming_message_receiver.recv().await {
+                match message {
+                    Message::NetworkData(network_data) => {
+                        // if we were to await a send the relay loop may get stuck if we are not
+                        // running a session
+                        network_data_sender.try_send(network_data).ok();
+                    }
+                    Message::BlockRequest(index) => {
+                        if let Some(signed_block) = db::load_block(&db_clone, index).await {
+                            sender_clone
+                                .send((Message::Block(signed_block), Recipient::Peer(peer_id)))
+                                .await
+                                .ok();
+
+                            tracing::info!(
+                                "Served the block with index {} to peer {}",
+                                index,
+                                peer_id
+                            );
+                        }
+                    }
+                    Message::Block(signed_block) => {
+                        // if we were to await a send the relay loop may get stuck if we are not
+                        // running a session
+                        signed_block_sender.try_send(signed_block).ok();
+                    }
                 }
             }
-            Message::Block(signed_block) => {
-                signed_block_sender.send(signed_block).await.ok();
-            }
+
+            std::future::pending().await
+        });
+
+        Self {
+            keychain,
+            db,
+            mempool_item_receiver,
+            outgoing_message_sender,
+            network_data_receiver,
+            signed_block_receiver,
+            relay_handle,
         }
     }
 
-    std::future::pending().await
-}
+    /// The receiver returns a sequence of items which is a subsequence of
+    /// all items ordered in this session and a supersequence of the accepted
+    /// items. The end of a session is signaled by the return of Some(None)
+    /// while the return of None directly signals that the session has been
+    /// interrupted, either by a call to shutdown or by dropping a
+    /// decision_sender without sending a decision.
+    pub async fn run_session(
+        &self,
+        index: u64,
+    ) -> mpsc::Receiver<Option<(OrderedItem, oneshot::Sender<Decision>)>> {
+        let (ordered_item_sender, ordered_item_receiver) = mpsc::channel(256);
```

Comment:
1. What happens if this channel fills up? Is there any risk of getting stuck, thus the high limit?
2. This seems to allow AlephBFT to be 256 items ahead of Fedimint consensus, which might lead to a lot of outdated contributions (for the ones that depend on previous consensus items).

This channel size seems like something we want to expose for Fedimint to configure, much like [`max_future_epochs` in HBBFT](https://docs.rs/hbbft/0.1.1/hbbft/honey_badger/struct.Params.html#structfield.max_future_epochs).

Reply:
No there is no risk of getting stuck. The reason this channel is not of capacity 1 is to prevent switching tasks for every item under load. So this value is somewhat random, we could make it smaller. Any specific suggestions?

Reply:
Makes sense. How many items do we expect at most per second right now? That would sound like a reasonable number (task switching cost is relatively low compared to threads and even threads do ok with time slices on the order of a few 100ms).

Not urgent, can be fixed later.